### PR TITLE
Fixed regression for SettingWithCopyWarning showing incorrect stacklevel

### LINKED ISSUE #42570: BUG: Less helpful `SettingWithCopyWarning` on new pandas version
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
# Your code here
import numpy as np
import pandas as pd

df = pd.DataFrame(np.arange(25).reshape(5,5))

def valid_indexes(in_df: pd.DataFrame):
  return in_df.loc[:3]

valid = valid_indexes(df)
valid[2] = 3
```

#### Problem description
Before (pandas 1.2.*), the resulting warning showed my own code:

```
See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
valid[2] = 3
```
Now, it only shows library code:
```
See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._set_item(key, value)
```

This is significantly harder to diagnose.

#### Expected Output
The line of code I wrote to produce the error

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : f00ed8f47020034e752baf0250483053340971b0
python           : 3.9.4.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 60 Stepping 3, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252
pandas           : 1.3.0
numpy            : 1.20.3
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.1.3
setuptools       : 52.0.0.post20210125
Cython           : None
pytest           : 6.2.4
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 1.4.3
lxml.etree       : 4.6.3
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.1
IPython          : 7.23.1
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.2
numexpr          : None
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.3
sqlalchemy       : 1.4.15
tables           : None
tabulate         : None
xarray           : None
xlrd             : 2.0.1
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #42570
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Is there a better way to check the stacklevel here?

### CODE REVIEW

File: pandas/tests/indexing/test_chaining_and_caching.py [hunk 1]
```diff
@@ -435,6 +435,16 @@ def test_detect_chained_assignment_warnings_filter_and_dupe_cols(self):
             )
             tm.assert_frame_equal(df, expected)
 
+    @pytest.mark.parametrize("rhs", [3, DataFrame({0: [1, 2, 3, 4]})])
+    def test_detect_chained_assignment_warning_stacklevel(self, rhs):
+        # GH#42570
+        df = DataFrame(np.arange(25).reshape(5, 5))
+        chained = df.loc[:3]
+        with option_context("chained_assignment", "warn"):
+            with tm.assert_produces_warning(com.SettingWithCopyWarning) as t:
```
Comment (jreback) [inline pandas/tests/indexing/test_chaining_and_caching.py:444]: i think `check_stacklevel` is *supposed* to do this?

File: pandas/core/frame.py [hunk 2]
```diff
@@ -3745,7 +3745,7 @@ def _set_item_mgr(self, key, value: ArrayLike) -> None:
         # try to set first as we want an invalid
         # value exception to occur first
         if len(self):
-            self._check_setitem_copy()
+            self._check_setitem_copy(stacklevel=5)
```
Comment (jreback) [inline pandas/core/frame.py:3748]: you can use the `find_stack_level` function i think its called

Comment (jbrockmendel) [reply pandas/core/frame.py:3748]: yah at first i was trying to use find_stack_level sparingly, but might as well go for it

File: pandas/tests/indexing/test_chaining_and_caching.py (hunk 1 above)
Comment (phofl) [reply pandas/tests/indexing/test_chaining_and_caching.py:444]: Yeah thought so too, but test was passing with stacklevel 4 and 5

File: pandas/core/frame.py (hunk 2 above)
Comment (phofl) [reply pandas/core/frame.py:3748]: Will do that as a follow up, since this would change the function signature I don't think we should backport this?

Comment (jreback) [reply pandas/core/frame.py:3748]: sure that's fine

**Review by jreback — APPROVED** [review]
lgtm

**Comment by phofl** [comment]
merging

**Pull request merged**